src/Microsoft.AspNet.Http/DefaultHttpResponse.cs

```diff
@@ -41,9 +43,18 @@ public override int StatusCode
             set { HttpResponseFeature.StatusCode = value; }
         }
 
-        public override IHeaderDictionary Headers
+        public override IDictionary<string, StringValues> Headers
         {
-            get { return new HeaderDictionary(HttpResponseFeature.Headers); }
+            get
+            {
+                if (HttpResponseFeature.Headers is SafeDictionary<string, StringValues>)
```

Comment:
Looked at using a local struct wrapper; but the Header collection is RO here, but R/W through the feature; so without a notification on the feature that its been changed it has to be something like this; and SafeDictionary a class as it will get boxed so no saving.

Headers Dict can be made R/W here safely.